### ISSUE TITLE
'prefer-const' not working correctly with 'for of' loop and generator-function.

### ISSUE THREAD

**Issue by virtyaluk** [eslint/eslint#6337]
**What version of ESLint are you using?**
2.10.2

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**
[.eslintrc](https://github.com/paper-ui/react-paper-ui/blob/master/.eslintrc)

**What did you do? Please include the actual source code causing the issue.**
```js
const foo = function *() {
    yield 1;
    yield 2;
    yield 3;
    yield 4;

    return 5;
};

for (let v of foo()) {
    console.log(v);
}

// which will return
// 1
// 2
// 3
// 4
```

I may repalce it with `for (const v of foo())`, but then `console.log` returns `1` for 4 times.

**What did you expect to happen?**
 To not treat `v` as a `const` while it gets a new value each iteration

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
ESLint: 'v' is never reassigned, use 'const' instead. (prefer-const)
```

Thanks!

**Comment by virtyaluk** [comment]
It may refer to #5568 indirectly.

**Comment by mysticatea** [comment]
Thank you for this issue.

This is duplicate of #5568, #4933, #4838, #4570, #4333, #4088, and #3104.
This is correct behavior.
Just some platform (e.g. Node 4, 5, Firefox) have a bug (see also https://github.com/eslint/eslint/issues/3104#issuecomment-157743822).

If you are using Node, you may be interested in [node/no-unsupported-features](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-unsupported-features.md) rule.

**Comment by virtyaluk** [comment]
Thanks!

I'm reading the specification and haven't finished it yet. So I can be wrong about different implementation pitfalls. And yeah, I'm testing it using Node because it's the fastest way to do that.

**Comment by mysticatea** [comment]
Node 4.x and 5.x does not support `let` and `const` completely in sloppy mode :smiley: 
Please try `"use strict"`.
